packages/arb-token-bridge-ui/src/hooks/useRedeemTeleporter.ts

```diff
@@ -0,0 +1,150 @@
+import { L1ToL2MessageData, L2ToL3MessageData } from './useTransactions'
+
+export type UseRedeemRetryableResult = {
+  redeem: () => Promise<void>
+  isRedeeming: boolean
+}
+
+export function useRedeemTeleporter(
+  tx: MergedTransaction,
+  address: Address | undefined
+): UseRedeemRetryableResult {
+  const chainIdForRedeemingRetryable = getChainIdForRedeemingRetryable(tx)
+
+  const { data: signer } = useSigner({
+    chainId: chainIdForRedeemingRetryable
+  })
+  const { updatePendingTransaction } = useTransactionHistory(address)
+
+  const redeemerNetworkName = getNetworkName(chainIdForRedeemingRetryable)
+
+  const [isRedeeming, setIsRedeeming] = useState(false)
+
+  const redeem = useCallback(async () => {
+    if (isRedeeming) {
+      return
+    }
+    try {
+      setIsRedeeming(true)
+
+      if (!signer) {
+        throw 'Signer is undefined'
+      }
+
+      // 1. first check which leg of the tx needs redeeming - l1toL2 or l2tol3
+      // 2. then, ensure that the user is connected to the proper providers and signers
+      // 3. then, we can call the same redeem logic we have on any of the Retryables and update the tx accordingly
+
+      const isFirstRetryableBeingRedeemed = firstRetryableRequiresRedeem(tx)
+
+      let sourceChainTxHash, retryableCreationId, sourceChainProvider
+
+      if (isFirstRetryableBeingRedeemed) {
+        sourceChainTxHash = tx.txId
+        retryableCreationId = tx.l1ToL2MsgData?.retryableCreationTxID
+        sourceChainProvider = getProviderForChainId(tx.parentChainId)
+      } else {
+        if (tx.l2ToL3MsgData) {
+          sourceChainTxHash = tx.l1ToL2MsgData?.l2TxID
+          retryableCreationId = tx.l2ToL3MsgData.retryableCreationTxID
+          sourceChainProvider = getProviderForChainId(
+            tx.l2ToL3MsgData.l2ChainId
+          )
+        }
+      }
+
+      if (!sourceChainTxHash || !sourceChainProvider) {
+        throw 'Could not find redemption details'
+      }
+
+      const retryableTicket = await getRetryableTicket({
+        parentChainTxHash: sourceChainTxHash,
+        retryableCreationId,
+        parentChainProvider: sourceChainProvider,
+        childChainSigner: signer
+      })
+
+      const reedemTx = await retryableTicket.redeem({ gasLimit: 40_000_000 }) // after a few trials, this gas limit seems to be working fine
+      await reedemTx.wait()
+
+      const status = await retryableTicket.status()
+      const isSuccess = status === L1ToL2MessageStatus.REDEEMED
+
+      if (!isSuccess) {
+        console.error('Redemption failed; status is not REDEEMED', reedemTx)
+        throw new Error(
+          'Redemption failed; status is not REDEEMED. Please try again later.'
+        )
+      }
+
+      const redeemReceipt = (await retryableTicket.getSuccessfulRedeem()) as {
+        status: L1ToL2MessageStatus.REDEEMED
+        l2TxReceipt: TransactionReceipt
+      }
+
+      const updatesInTx = isFirstRetryableBeingRedeemed
+        ? {
```

Comment:
This is quite difficult to follow, it's readable but I feel like it could be cleaned up and abstracted more

Reply:
I have separated the conditionals, maybe it's easier to follow now?

Reply:
yep nice, also `txUpdateCallback` uses await but it's not an async function

Reply:
Updated